### ISSUE TITLE
Can't assign a function with a NoReturn return type to a callable with matching signature

### ISSUE THREAD

**Issue by FasterSpeeding** [microsoft/pyright#2104]
**Describe the bug**
Attempts to assign a callback with the bottom return type of NoReturn to a field with matching argument signature but a real return type leads to pyright erroring on NoReturn being incompatible with the real return type.
(As a note this can also lead to errors when overriding a method with a NoReturn return type if reportIncompatibleMethodOverride is set to error level).

**To Reproduce**

* Lint the example code with pyright
* Watch pyright output the relevant error

```py
  C:\...\typing_test.py:12:15 - error: Argument of type "() -> NoReturn" cannot be assigned to parameter "callback" of type "() -> bool" in
function "add_rule"
    Type "() -> NoReturn" cannot be assigned to type "() -> bool"
      Function return type "NoReturn" is incompatible with type "bool"
        "NoReturn" is incompatible with "bool" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
Completed in 0.54sec

```

**Expected behavior**
Pyright should allow what is just narrowing the return type from one or multiple types to the lack of a return type (since NoReturn is a bottom type). Since this behaviour is helpful for cases such as when a specific implementation of a method will always error (hence NoReturn being fitting for that impl) or for more accurately typing callbacks which by design always raise but are passed in places where the callback is expected to either raise or return.

**Screenshots or Code**

```python
import typing


def raise_unexpected() -> typing.NoReturn:
    raise RuntimeError("Unexpected case")


def add_rule(type_: str, callback: typing.Callable[[], bool], /) -> None:
    ...


add_rule("*", raise_unexpected)
```

**VS Code extension or command-line**
pyright 1.1.157 (command-line tool)
targetting python 3.9

**Additional context**

**Comment by erictraut** [comment]
I don't fully agree with your argument, but I understand it and see how this exception would be useful. Since mypy accepts the above code, it makes sense for pyright to do the same.

**Comment by erictraut** [comment]
This is included in pyright 1.1.158, which I just published. It will also be included in the next release of pylance.
